protocol/x/clob/e2e/long_term_orders_test.go

```diff
@@ -69,11 +71,85 @@ func TestPlaceOrder_StatefulCancelFollowedByPlaceInSameBlockErrorsInCheckTx(t *t
 	require.NotContains(t, orders, LongTermPlaceOrder_Alice_Num0_Id0_Clob0_Buy5_Price10_GTBT5.Order.OrderId)
 }
 
+// TestCancelFullyFilledStatefulOrderInSameBlockItIsFilled tests the scenario where an honest block proposer
+// may propose a stateful cancellation which fails because the order was fully filled in the same block.
+func TestCancelFullyFilledStatefulOrderInSameBlockItIsFilled(t *testing.T) {
+	tApp := testapp.NewTestAppBuilder(t).Build()
+	ctx := tApp.InitChain()
+
+	// Place order
+	result := tApp.CheckTx(testapp.MustMakeCheckTx(
+		ctx,
+		tApp.App,
+		testapp.MustMakeCheckTxOptions{
+			AccAddressForSigning: testtx.MustGetOnlySignerAddress(&LongTermPlaceOrder_Alice_Num0_Id0_Clob0_Buy5_Price10_GTBT5),
+		},
+		&LongTermPlaceOrder_Alice_Num0_Id0_Clob0_Buy5_Price10_GTBT5,
+	))
+	require.True(t, result.IsOK(), "Expected CheckTx to succeed. Response: %+v", result)
+	ctx = tApp.AdvanceToBlock(2, testapp.AdvanceToBlockOptions{})
+
+	// Place order which fully matches the first order
+	result = tApp.CheckTx(testapp.MustMakeCheckTx(
+		ctx,
+		tApp.App,
+		testapp.MustMakeCheckTxOptions{
+			AccAddressForSigning: testtx.MustGetOnlySignerAddress(&PlaceOrder_Bob_Num0_Id0_Clob0_Sell5_Price10_GTB20),
+		},
+		&PlaceOrder_Bob_Num0_Id0_Clob0_Sell5_Price10_GTB20,
+	))
+	require.True(t, result.IsOK(), "Expected CheckTx to succeed. Response: %+v", result)
+
+	// Place cancellation
+	cancellationTx := testapp.MustMakeCheckTx(
+		ctx,
+		tApp.App,
+		testapp.MustMakeCheckTxOptions{
+			AccAddressForSigning: testtx.MustGetOnlySignerAddress(&constants.CancelLongTermOrder_Alice_Num0_Id0_Clob0_GTBT15),
+		},
+		&constants.CancelLongTermOrder_Alice_Num0_Id0_Clob0_GTBT15,
+	)
+	result = tApp.CheckTx(cancellationTx)
+	require.True(t, result.IsOK(), "Expected CheckTx to succeed. Response: %+v", result)
+
+	// DeliverTx should fail for cancellation tx
+	ctx = tApp.AdvanceToBlock(3, testapp.AdvanceToBlockOptions{
+		ValidateDeliverTxs: func(
+			ctx sdktypes.Context,
+			request abcitypes.RequestDeliverTx,
+			response abcitypes.ResponseDeliverTx,
+			txIndex int,
+		) (haltChain bool) {
+			if lib.SlicesAreEqual(request.Tx, cancellationTx.Tx) {
```

Comment:
these are just byte arrays, right?

we can just use bytes.Equal. I think it's slightly more optimized and we don't have to write tests for `SlicesAreEqual`

Reply:
maybe we can also use `txIndex`, which might be easier

Reply:
oof, didn't know about bytes.Equal. Yeah I was thinking of using txIndex, I guess I can use that. Should reliably be able to get the same txIndex for the transaction I'm looking for